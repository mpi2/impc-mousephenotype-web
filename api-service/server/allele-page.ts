import { fetchAPIFromServer, fetchURL } from "@/api-service";
import { AlleleSummary } from "@/models";
const KUBERNETES_NAMESPACE = process.env.KUBERNETES_NAMESPACE ?? "default";
const NODE_ENV = process.env.NODE_ENV;

export async function fetchAlleleSummary(
  mgiGeneAccessionId: string,
  alleleName: string,
): Promise<AlleleSummary> {
  const endpointURL = `http://impc-allele-service.${KUBERNETES_NAMESPACE}:8080/v1/alleles?mgiGeneAccessionId=${mgiGeneAccessionId}&alleleName=${alleleName}`;
  return await (NODE_ENV === "development"
    ? fetchAPIFromServer<AlleleSummary>(
        `/api/v1/alleles/${mgiGeneAccessionId}/${alleleName}`,
      )
    : fetchURL<AlleleSummary>(endpointURL));
}
