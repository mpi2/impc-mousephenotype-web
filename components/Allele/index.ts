export { default as AlleleMap } from './AlleleMap';
export { default as Crispr } from './Crispr';
export { default as ESCell } from './ESCell';
export { default as IVP } from './IVP';
export { default as Mice } from './Mice';
export { default as QCModal } from './QCModal';
export { default as TVP } from './TVP';
