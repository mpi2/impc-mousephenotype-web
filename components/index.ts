export { default as Card } from './Card';
export { default as Search } from './Search';
export { default as AlleleSymbol } from './AlleleSymbol';
export { default as DownloadData } from './DownloadData'