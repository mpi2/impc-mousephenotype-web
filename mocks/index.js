export async function initMocks() {
  if (typeof window === "undefined") {
    const { server } = await import("./server");
    server.listen();
  } else {
    const { worker } = await import("./browser");
    return worker.start({
      serviceWorker: {
        url: "/data/mockServiceWorker.js",
        options: {
          scope: "/",
        }
      }
    });
  }
}

