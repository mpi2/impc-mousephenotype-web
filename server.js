const { createServer } = require('http');
const { parse } = require('url');
const next = require('next');

const dev = process.env.NODE_ENV !== 'production';
const hostname = 'localhost';
const port = 3000;
const app = next({ dev, hostname, port });
const handle = app.getRequestHandler();

app.prepare()
.then(() => {
    createServer(async (req, res) => {
        try {
            const parsedUrl = parse(req.url, true);
            const { pathname, query} = parsedUrl;
            const host = req.headers.host;
            console.log({ host, pathname });
            res.setHeader('X-IMPC-TEST', "true");
            if (host.includes('nginx.mousephenotype-dev.org')) {
                const modifiedPathName = pathname.replace("/data/", "");
                console.log({ modifiedPathName });
                await app.render(req, res, modifiedPathName, query);
                res.setHeader('X-PATH', JSON.stringify({ host, pathname, modifiedPathName }));
            } else {
                await handle(req, res, parsedUrl);
            }

        } catch (err) {
            console.error('Error occurred handling', req.url, err)
            res.statusCode = 500
            res.end('internal server error')
        }
    })
        .once('error', (err) => {
            console.error(err);
            process.exit(1);
        })
        .listen(port, () => {
            console.log(`Ready on http://${hostname}:${port}`);
        })
});