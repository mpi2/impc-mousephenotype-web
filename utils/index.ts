import _ from "lodash";

export const formatBodySystems = (systems: string[] | string = []) => {
  return _.capitalize(
    (typeof systems === "string" ? systems : systems.join(", "))
      .replace(/ phenotype/g, "")
      .replace(/\//g, " / ")
  );
};

export const formatAlleleSymbol = (allele: string) => {
  return allele.slice(0, allele.length - 1).split("<");
};
