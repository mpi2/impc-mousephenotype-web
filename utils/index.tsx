import React from "react";
import _ from "lodash";

export const formatBodySystems = (systems: string[] | string = []) => {
  return _.capitalize(
    (typeof systems === "string" ? systems : systems.join(", "))
      .replace(/ phenotype/g, "")
      .replace(/\//g, " / ")
  );
};

export const formatAlleleSymbol = (allele: string) => {
  return allele.slice(0, allele.length - 1).split("<");
};

export const formatPValue = (pValue: number) => {
  const pValueArray = Number.parseFloat(String(pValue))
    .toExponential(2)
    .split("e");
  return (
    <>
      {pValueArray[0]}x10<sup>{pValueArray[1].replace("+", "")}</sup>
    </>
  );
};

export const formatESCellName = (src: string) => {
  const [name, superscript, end] = src.split(/<|>/);
  return (
    <>
      {name}
      <sup>{superscript}</sup>
      {end}
    </>
  );
};

const fetchCache = {};
export { fetchCache };

export const toSentenceCase = (camelCase) => {
  if (camelCase) {
    const result = camelCase.replace(/([A-Z])/g, " $1");
    return result[0].toUpperCase() + result.substring(1).toLowerCase();
  }
  return "";
};
