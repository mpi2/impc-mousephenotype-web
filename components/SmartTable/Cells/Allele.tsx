import { Model, TableCellProps } from "@/models";
import _ from "lodash";
import { formatAlleleSymbol } from "@/utils";

const AlleleCell = <T extends Model>(props: TableCellProps<T>) => {
  const allele = formatAlleleSymbol(_.get(props.value, props.field) as string);
  return (
    <>
      {allele[0]}
      <sup>{allele[1]}</sup>
    </>
  )
};

export default AlleleCell;