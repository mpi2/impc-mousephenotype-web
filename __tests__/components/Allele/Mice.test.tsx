import { screen, waitFor } from '@testing-library/react';
import { Mice } from "@/components/Allele";
import { API_URL, renderWithClient } from "../../utils";
import mockRouter from "next-router-mock";
import { server } from "../../../mocks/server";
import { rest } from "msw";
import dbn1Mice from "../../../mocks/data/tests/dbn1-mice.json";
import userEvent from '@testing-library/user-event';

jest.mock('next/router', () => jest.requireActual('next-router-mock'));

describe('Allele Page - Mice', () => {
  beforeEach(async () => {
    const endpointURL = `${API_URL}/api/v1/alleles/mice/get_by_mgi_and_allele_name/MGI:1931838/tm1a(KOMP)Wtsi`;
    server.use(
      rest.get(endpointURL, (req, res, ctx) => {
        return res(ctx.status(200), ctx.json(dbn1Mice));
      })
    );

    await mockRouter.push('/data/alleles/MGI:1931838/tm1a(KOMP)Wtsi');
  });
  it('renders table with 5 columns when is Crispr', async () => {
    const user = userEvent.setup();

    const qcDataFn = jest.fn();

    const { container} = renderWithClient(
      <Mice
        mgiGeneAccessionId="MGI:1931838"
        alleleName="tm1a(KOMP)Wtsi"
        isCrispr
        setQcData={qcDataFn}
      />
    );
    await waitFor(() => expect(screen.getByRole('table')).toBeInTheDocument());
    expect(container).toMatchSnapshot();

    await user.click(screen.getByRole('link', { name: "View" }));
    expect(qcDataFn).toHaveBeenCalledWith([{
      "productionQc": {
        "tvBackboneAssay": "pass",
        "fivePrimeCassetteIntegrity": "pass",
        "neoCountQpcr": "pass",
        "loaQpcr": "pass",
        "laczSrPcr": "pass",
        "mutantSpecificSrPcr": "pass",
        "loxpConfirmation": "pass",
        "fivePrimeLrPcr": "fail",
        "homozygousLoaSrPcr": "pass",
        "threePrimeLrPcr": "pass"
      }
    }]);
    expect(screen.getAllByRole('columnheader').length).toBe(5);
  });

  it('renders table with 6 columns when is an ESCell', async () => {
    const qcDataFn = jest.fn();

    renderWithClient(
      <Mice
        mgiGeneAccessionId="MGI:1931838"
        alleleName="tm1a(KOMP)Wtsi"
        isCrispr={false}
        setQcData={qcDataFn}
      />
    );
    await waitFor(() => expect(screen.getByRole('table')).toBeInTheDocument());
    expect(screen.getAllByRole('columnheader').length).toBe(6);
  });

  it('displays its empty state when there is no data', async() => {
    const qcDataFn = jest.fn();

    const endpointURL = `${API_URL}/api/v1/alleles/mice/get_by_mgi_and_allele_name/MGI:1931838/tm1a(KOMP)WtsiTest`;
    server.use(
      rest.get(endpointURL, (req, res, ctx) => {
        return res(ctx.status(200), ctx.json([]));
      })
    );

    renderWithClient(
      <Mice
        mgiGeneAccessionId="MGI:1931838"
        alleleName="tm1a(KOMP)WtsiTest"
        isCrispr={false}
        setQcData={qcDataFn}
      />
    );
    await waitFor(() => expect(screen.getByTestId('no-results-alert')).toBeDefined());
  });
});