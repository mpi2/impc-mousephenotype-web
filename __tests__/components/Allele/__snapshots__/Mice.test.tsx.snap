// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Allele Page - Mice renders table with 5 columns when is Crispr 1`] = `
<div>
  <div
    class="card"
    id="mice"
  >
    <h2
      data-testid="header"
    >
      Mice
    </h2>
    <div
      class="buttonsWrapper top "
      data-testid="top-controls-wrapper"
    />
    <div
      class="tableWrapper"
    >
      <table
        class="table striped undefined table table-bordered"
      >
        <thead>
          <tr>
            <th
              rowspan="2"
              width="25%"
            >
              <button
                class="inlineButton headerButton disabled"
                style="font-weight: normal;"
              >
                Colony Name
                 
              </button>
            </th>
            <th
              rowspan="2"
              width="16.666666666666664%"
            >
              <button
                class="inlineButton headerButton disabled"
                style="font-weight: normal;"
              >
                Genetic Background
                 
              </button>
            </th>
            <th
              rowspan="2"
              width="16.666666666666664%"
            >
              <button
                class="inlineButton headerButton disabled"
                style="font-weight: normal;"
              >
                Production Centre
                 
              </button>
            </th>
            <th
              rowspan="2"
              width="8.333333333333332%"
            >
              <button
                class="inlineButton headerButton disabled"
                style="font-weight: normal;"
              >
                QC Data
                 
              </button>
            </th>
            <th
              rowspan="2"
              width="16.666666666666664%"
            >
              <button
                class="inlineButton headerButton disabled"
                style="font-weight: normal;"
              >
                Order / Contact
                 
              </button>
            </th>
          </tr>
          <tr />
        </thead>
        <tbody>
          <tr>
            <td>
              <strong>
                MCKS
              </strong>
            </td>
            <td>
              C57BL/6NTac
            </td>
            <td>
              WTSI
            </td>
            <td>
              <a
                class="link"
                href="#"
                target="_blank"
              >
                View
                 
                <svg
                  aria-hidden="true"
                  class="svg-inline--fa fa-window-maximize fa-xs grey"
                  data-icon="window-maximize"
                  data-prefix="far"
                  focusable="false"
                  role="img"
                  viewBox="0 0 512 512"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M.3 89.5C.1 91.6 0 93.8 0 96V224 416c0 35.3 28.7 64 64 64l384 0c35.3 0 64-28.7 64-64V224 96c0-35.3-28.7-64-64-64H64c-2.2 0-4.4 .1-6.5 .3c-9.2 .9-17.8 3.8-25.5 8.2C21.8 46.5 13.4 55.1 7.7 65.5c-3.9 7.3-6.5 15.4-7.4 24zM48 224H464l0 192c0 8.8-7.2 16-16 16L64 432c-8.8 0-16-7.2-16-16l0-192z"
                    fill="currentColor"
                  />
                </svg>
              </a>
            </td>
            <td>
              <a
                class="link primary"
                href="https://www.infrafrontier.eu/emma/strain-search/?category=genesalleles&keyword=Dbn1"
                target="_blank"
              >
                <svg
                  aria-hidden="true"
                  class="svg-inline--fa fa-cart-shopping "
                  data-icon="cart-shopping"
                  data-prefix="fas"
                  focusable="false"
                  role="img"
                  viewBox="0 0 576 512"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M0 24C0 10.7 10.7 0 24 0H69.5c22 0 41.5 12.8 50.6 32h411c26.3 0 45.5 25 38.6 50.4l-41 152.3c-8.5 31.4-37 53.3-69.5 53.3H170.7l5.4 28.5c2.2 11.3 12.1 19.5 23.6 19.5H488c13.3 0 24 10.7 24 24s-10.7 24-24 24H199.7c-34.6 0-64.3-24.6-70.7-58.5L77.4 54.5c-.7-3.8-4-6.5-7.9-6.5H24C10.7 48 0 37.3 0 24zM128 464a48 48 0 1 1 96 0 48 48 0 1 1 -96 0zm336-48a48 48 0 1 1 0 96 48 48 0 1 1 0-96z"
                    fill="currentColor"
                  />
                </svg>
                 
                EMMA
              </a>
            </td>
          </tr>
        </tbody>
      </table>
    </div>
    <div
      class="buttonsWrapper withControls"
    >
      <div
        class="additionalWrapper bottomControls"
      >
        <div />
      </div>
    </div>
  </div>
</div>
`;
